### ISSUE TITLE
Centerpoint Time To Train

### ISSUE THREAD

**Issue by chucktang98** [open-mmlab/mmdetection3d#1318]
How long does an average epoch of centerpoint take? For some reason on my 8 GPU set up it takes > 5+ days to train which seems abnormal? Does that mean my ground_truth_database was incorrectly generated? What is the time metric in the training logs [here](https://download.openmmlab.com/mmdetection3d/v0.1.0_models/pointpillars/hv_pointpillars_secfpn_sbn-all_4x8_2x_nus-3d/hv_pointpillars_secfpn_sbn-all_4x8_2x_nus-3d_20200620_230725.log.json)?

**Comment by ZCMax** [comment]
ground_truth_database is generated before training, and used for data augmentation. Can you provide your training config, and did you train on waymo dataset?

**Comment by chucktang98** [comment]
Here's my [config](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/centerpoint/centerpoint_01voxel_second_secfpn_dcn_circlenms_4x8_cyclic_20e_nus.py) and the logs below. I didn't try training on the waymo dataset. Can you clarify what the time unit is in the log below (is it minutes?). It seems like a lot of the time is spent moving the data around since my gt_database is stored in the NAS and not locally, maybe I should try moving it to local and see how that does? 

```
HardDiskBackend.
/opt/conda/lib/python3.7/site-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /opt/conda/conda-bld/pytorch_1623448265233/work/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
2022-03-17 08:40:09,828 - mmcv - INFO - Reducer buckets have been rebuilt in this iteration.
2022-03-17 09:01:02,704 - mmdet - INFO - Epoch [1][50/123580]   lr: 1.000e-04, eta: 753 days, 19:17:14, time: 26.351, data_time: 25.998, memory: 1424, task0.loss_heatmap: 340.4277, task0.loss_bbox: 1.4971, task1.loss_heatmap: 297.1169, task1.loss_bbox: 1.6756, task2.loss_heatmap: 662.3008, task2.loss_bbox: 1.9460, task3.loss_heatmap: 512.4426, task3.loss_bbox: 1.4658, task4.loss_heatmap: 346.4776, task4.loss_bbox: 1.1984, task5.loss_heatmap: 520.5218, task5.loss_bbox: 1.3682, loss: 2688.4385, grad_norm: 17938.6001
```

**Comment by chucktang98** [comment]
Also do you have a link of the `gt_database` stored in s3 or google drive by any chance? Curious if my `gt_database` was generated incorrectly?

**Comment by Tai-Wang** [comment]
First, it seems your machine does not have many GPUs? We recommend you try nuScenes experiments with at least 4 GPUs or 8 GPUs because its samples are much more than KITTI. Besides, the data_time can include many aspects, not only reading the gt_database. You need to have a closer look into the data reading overhead for each stage (such as reading raw data, data info and gt_database for augmentation).

**Comment by chucktang98** [comment]
I modified the `train.py` script so I could test on just 1 GPU :-). I also have a cloud version running on determined that has 8 GPU's that runs into the same issue as above. For the data issue, I think something is wrong with my gt_database since I moved all my data to disk as opposed to the NAS and the data_time is still 25 instead of 1. Wondering if you could link me to a gt_database from s3 that I could try to download?

**Comment by chucktang98** [comment]
If the s3 bucket is too difficult to provide, wondering if you could send me some python jupyter scripts that show like for instance the length of the `nuscenes_db_infos_train.pkl` and maybe `du /media/syndata/nuscenes/nuscenes_gt_database -BG 1`?

**Comment by ZCMax** [comment]
We do have s3 data for internal use, but it's not open. You can check the `nuscenes_dbinfos_train.pkl` for easy check:
```
ann_file = 'data/nuscenes/nuscenes_dbinfos_train.pkl'
data = mmcv.load(ann_file)
for key, value in data.items():
    print(key+ ' num :' + str(len(value)))
```
The results should be:
```
truck num :65262
pedestrian num :161928
car num :339949
movable_object.debris num :2120
traffic_cone num :62964
motorcycle num :8846
static_object.bicycle_rack num :2259
movable_object.pushable_pullable num :19195
vehicle.emergency.ambulance num :11
construction_vehicle num :11050
trailer num :19202
barrier num :107507
bicycle num :8185
bus num :12286
vehicle.emergency.police num :498
human.pedestrian.stroller num :751
animal num :619
human.pedestrian.wheelchair num :492
human.pedestrian.personal_mobility num :352
```

**Comment by chucktang98** [comment]
Got it and the data time is in seconds right for the logs? Curious how long it takes for you guys to train 1 centerpoint model on 8/4 gpus?

**Comment by chucktang98** [comment]
@ZCMax our datasets are the same so there is no problem there. I think the issue is in our distributed FileClient backend which is why only `ObjectSample` operation is extremely slow. And I am assuming that it takes around 1 hour to train with 8 GPUs from reading your logs, is that correct. I am trying to reproduce the baseline and am running into around a 9 hour training time right now (with evaluation every epoch).

**Comment by chucktang98** [comment]
@ZCMax @Tai-Wang I think the `data_time` parameter is in seconds from looking at the code in [iter_timer.py](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/iter_timer.py). This means based on these calculations, it takes: `3 seconds / 50 iterations * 3850 iterations * 20 (epochs) =  4620 seconds` or 1 hour to train a centerpoint [model](https://github.com/open-mmlab/mmdetection3d/blob/master/configs/centerpoint/centerpoint_01voxel_second_secfpn_dcn_circlenms_4x8_cyclic_20e_nus.py). Please let me know if this is incorrect as I will be trying to get my training time down from 2 days in our determined cloud to 1 hour in the next few days. Also, curious where you would think my off by 60 factor is and how you would go about debugging this?

One more question on this, if your data is training at 50 * 4 (batch size) per 3 seconds then your frames per second is 60 fps for a single stage centerpoint! Is this correct as it does seem a bit fast?

**Comment by chucktang98** [comment]
Ah I see, the `iter_timer.py` only records the time it takes for every iteration and is only logged every 50 iterations. That's why the total time is actually  3 seconds / iteration *  50 iterations * 3850 iterations * 20 (epochs) = 64 hours or 2.5 days to train which is similar to our training set up :-). This is also in line with the reported times described [here](https://github.com/tianweiy/CenterPoint/blob/master/configs/waymo/README.md)
